### PR TITLE
Adding `ignored_parameters` feature

### PR DESCRIPTION
**Pull request by themiurgo**
This pull request adds the possibility to exclude some request parameters from the cache key. This is useful in some occasions, for example when requesting the same resources with more access tokens, specified as parameters.

Please let me know if this needs any change to be included in the original project.

### CODE REVIEW

**Comment by reclosedev** [comment]
I like the idea.

Code needs some modifications. There are 2 tests failing now, let me explain why.
``_normalize_params`` is needed because when you pass ``dict``, the order of parameters is undefined. But whe you pass key-value pairs, it means, that you know the order, e.g. some weird sites require exact order of parameters. Also, there can some raw data passed.
In your code, you try to always convert parameters to dict.
https://travis-ci.org/reclosedev/requests-cache/jobs/62084996#L243-L286

I think we should exclude ignored parameters only if ``ignored_parameters  is not None``. Not sure about converting key-value pairs (or ``OrderedDict``) to dict in that case, though.

Also, please add simple test, something like:
```python
def test_ignore_parameters(self):
    url = httpbin("get")
    ignored_param = "ignored"
    usual_param = "some"
    s = CachedSession(CACHE_NAME, CACHE_BACKEND, ignored_parameters=[ignored_param])

    params = {ignored_param: "1", usual_param: "1"}
    s.get(url, params=params)
    self.assertTrue(s.get(url, params=params).from_cache)

    params[ignored_param] = "new"
    self.assertTrue(s.get(url, params=params).from_cache)

    params[usual_param] = "new"
    self.assertFalse(s.get(url, params=params).from_cache)
```

**Comment by themiurgo** [comment]
Thanks for the tips. So, what kind of stuff can get passed as `params`? dict, OrderedDict and (key, item) lists?

File: requests_cache/core.py [hunk 1]
```diff
@@ -259,12 +264,18 @@ def clear():
...
+def _normalize_parameters(params, ignored_parameters=None):
     """ If builtin dict is passed as parameter, returns sorted list
     of key-value pairs
     """
-
     if type(params) is dict:
-        return sorted(params.items(), key=itemgetter(0))
+        params = sorted(params.items(), key=itemgetter(0))
+
+    if ignored_parameters:
```
Comment (reclosedev) [inline requests_cache/core.py]: If custom dict, like ``OrderedDict``, is passed, ``type(params) is dict`` is ``False`` (it's intentional) and ``for k, v in params`` will fail. So, please add check here:
``` python
if isinstance(params, collections.Mapping):
   params = params.items()
```

File: requests_cache/core.py [hunk 2]
```diff
@@ -259,12 +264,18 @@ def clear():
...
+        try:
+            params = [(key, value) for key, value in params \
```
Comment (reclosedev) [inline requests_cache/core.py]: Please remove ``\``, it's redundant between brackets. Also something wrong with indentation on the next line.

File: requests_cache/core.py [hunk 3]
```diff
@@ -259,12 +264,18 @@ def clear():
 def _patch_session_factory(session_factory=CachedSession):
     requests.Session = requests.sessions.Session = session_factory
 
-
```
Comment (reclosedev) [inline requests_cache/core.py]: New line removed accidentally.

**Comment by reclosedev** [comment]
Looks good! 
Please fix small issues in inline comments, and I'll merge.

**Comment by themiurgo** [comment]
Fixed! Let me know if this looks final and if you want the commits to be squashed before merging.

**Comment by reclosedev** [comment]
Thank you!

**Comment by themiurgo** [comment]
No problem, keep up the great work!

**Comment by YetAnotherNerd** [comment]
i might got it wrong, but shouldn't the parameters just be ignored for the sake of cachekeys and still be present in the real request that is going to be send?
i try to cache rauth requests, which seems not to work due to the changing oath_{timestamp,nonce,signature} params. ignoring them with the above code seems to make them get cached, but also make the requests invalid (since they are missing those params)

**Comment by themiurgo** [comment]
If the ignored parameters are oauth_{timestamp,nonce,signature} there are two cases:
- the resource is already cached, hence those parameters won't be needed to fetch the resource from the cache;
- the resource is not available and in that case the request will proceed as normal.

Are you saying that in the second part does not work and when there is a cache miss a request will be forwarded with the ignored_params stripped out?

**Comment by YetAnotherNerd** [comment]
> Are you saying that in the second part does not work and when there is a cache miss a request will be forwarded with the ignored_params stripped out?

yep, thats exactly how it appears to me.
since i wasnt able to debug it using wireshark, i tried hacking in some print statements into requests.Session code. as soon as i set ignored_params they wont appear anymore, even with an empty cache.

**Pull request merged**